File: src/main/java/ar/edu/unq/spring/persistence/dao/ItemDAO.java
package ar.edu.unq.spring.persistence.dao;

import ar.edu.unq.spring.modelo.Item;
import ar.edu.unq.spring.persistence.dto.ItemJPADTO;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.Set;

@Repository
public interface ItemDAO extends JpaRepository<ItemJPADTO, Long> {

    ItemJPADTO findTopByOrderByPesoDesc();

    @Query(
            "FROM Item i where i.peso  > :peso order by i.peso asc"
    )
    Set<ItemJPADTO> getMasPesados(@Param("peso") int peso);

    @Query(
            "from Item i "
                    + "where i.owner.vida < ?1 "
                    + "order by i.peso asc"
    )
    Set<ItemJPADTO> getItemsDePersonajesDebiles(int vida);
}

File: src/main/kotlin/ar/edu/unq/spring/persistence/dao/ItemDAO.kt
package ar.edu.unq.spring.persistence.dao

import ar.edu.unq.spring.persistence.dto.ItemJPADTO
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.CrudRepository

interface ItemDAO : CrudRepository<ItemJPADTO, Long> {

    fun findTopByOrderByPesoDesc(): ItemJPADTO

    @Query(
        "from Item i "
                + "where i.owner.vida < ?1 "
                + "order by i.peso asc"
    )
    fun getItemsDePersonajesDebiles(unaVida: Int): Collection<ItemJPADTO>


    @Query(
            "FROM Item i where i.peso  > ?1 order by i.peso asc"
    )
    fun getMasPesados(unValorDado: Int): Collection<ItemJPADTO>

}
